src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts

```diff
@@ -183,66 +176,58 @@ function addScriptTagsToDocument(resourceLoader: WebAssemblyResourceLoader) {
     const resourceType: WebAssemblyBootResourceType = 'dotnetjs';
-    const customSrc = resourceLoader.startOptions.loadBootResource(resourceType, dotnetJsResourceName, scriptElem.src, dotnetJsContentHash);
+    const customSrc = resourceLoader.startOptions.loadBootResource(resourceType, dotnetJsResourceName, src, dotnetJsContentHash);
     if (typeof (customSrc) === 'string') {
-      scriptElem.src = customSrc;
+      src = customSrc;
     } else if (customSrc) {
       // Since we must load this via a <script> tag, it's only valid to supply a URI (and not a Request, say)
       throw new Error(`For a ${resourceType} resource, custom loaders must supply a URI string.`);
     }
   }
 
-  document.body.appendChild(scriptElem);
+  return new Promise((resolve, reject) => {
+    // GOTCHA: remove this callback once we switch to ES6
+    // this is callback we have in CJS version of the runtime
+    globalThis.__onDotnetRuntimeLoaded = (createDotnetRuntime) => {
+      delete globalThis.__onDotnetRuntimeLoaded;
+      resolve(createDotnetRuntime);
+    };
+    // this tells webpack to do dynamic import at runtime, not at compile time
+    import(/* webpackIgnore: true */ src).then(({ default: createDotnetRuntime }) => {
+      if (createDotnetRuntime) {
+        // this runs when loaded module was ES6
+        delete globalThis.__onDotnetRuntimeLoaded;
+        resolve(createDotnetRuntime);
+      }
+    }, reject);
+  });
 }
 
-// Due to a strange behavior in macOS Catalina, we have to delay loading the WebAssembly files
-// until after it finishes evaluating a <script> element that assigns a value to window.Module.
-// This may be fixed in a later version of macOS/iOS, or even if not it may be possible to reduce
-// this to a smaller workaround.
-function addGlobalModuleScriptTagsToDocument(callback: () => void) {
-  const scriptElem = document.createElement('script');
-
-  // This pollutes global but is needed so it can be called from the script.
-  // The callback is put in the global scope so that it can be run after the script is loaded.
-  // onload cannot be used in this case for non-file scripts.
-  window['__wasmmodulecallback__'] = callback;
+async function createEmscriptenModuleInstance(resourceLoader: WebAssemblyResourceLoader, onReady: () => void, onError: (reason?: any) => void) {
 
-  // Note: Any updates to the following script will require updating the inline script hash if using CSP
-  scriptElem.text = 'var Module; window.__wasmmodulecallback__(); delete window.__wasmmodulecallback__;';
-
-  document.body.appendChild(scriptElem);
-}
-
-function createEmscriptenModuleInstance(resourceLoader: WebAssemblyResourceLoader, onReady: () => void, onError: (reason?: any) => void) {
+  const dotnetJsBeingLoaded = importDotnetJs(resourceLoader);
   const resources = resourceLoader.bootConfig.resources;
-  const module = (window['Module'] || {}) as typeof Module;
+  const moduleConfig = (window['Module'] || {}) as typeof Module;
```

Comment:
This is for your consideration if your users expect and use `window.Module`. From runtime perspective, you could stop poluting `window`.

Reply:
This is something that ideally we want to cleanup by default, but I do believe we need an option for backwards compatibility. We can flow this via blazorwebassemblyoptions and have it be opt-in (to ease out anyone using this from it)

Reply:
+1 on dropping the global by default and having an opt-in to bring it back.

My only concern would be about discoverability. If people upgrade an app and it starts failing with `Uncaught ReferenceError: Module is not defined`, it doesn't give them any clue that there's a switch to fix it. Not sure how to solve that.

Reply:
I'm interested to listen on that design conversation, when you guys have it. It's tough problem.

We solved similar problem by polluting global namespace with `getDotnetRuntime(runtimeId: number)`. Our debugger now attaches to id `0`, assuming that there is at least one runtime on the page and that's the one which should be debugged.

This function could be also used to locate existing runtime and use it, instead of creating another instance on the page. If you have multiple WebComponents which don't know about each other, for example. Runtime has version exposed, so that you could know if it's compatible with you. But it doesn't tell you what AOT stripped.